File: 04-stack/00739-Daily_Temperatures/00739-daily_temperatures.cpp
// 739: Daily Temperstures
// https://leetcode.com/problems/daily-temperatures/


#include <iostream>
#include <vector>
#include <stack>

using namespace std;


class Solution {
public:
    // SOLUTION
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<int> s;
        vector<int> result(temperatures.size());

        for (int i=0; i<temperatures.size(); i++) {
            while (!s.empty() && temperatures[i]>temperatures[s.top()]) {
                result[s.top()] = i - s.top();
                s.pop();
            }
            s.push(i);
        }

        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> temperatures = {73,74,75,71,69,72,76,73};

    // OUTPUT
    auto result = o.dailyTemperatures(temperatures);
    cout<<"["; for (auto v : result) cout<<v<<" "; cout<<"\b]"<<endl;

    return 0;
}

File: 04-stack/00739-Daily_Temperatures/00739-daily_temperatures.py
# 739: Daily Temperstures
# https://leetcode.com/problems/daily-temperatures/


class Solution:
    # SOLUTION
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        s = []
        result = [0]*len(temperatures)

        for i in range(len(temperatures)):
            while len(s)!=0 and temperatures[i]>temperatures[s[-1]]:
                result[s[-1]] = i - s[-1]
                s.pop()
            s.append(i)

        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    temperatures: list[int] = [73,74,75,71,69,72,76,73]

    # OUTPUT
    result = o.dailyTemperatures(temperatures)
    print(result)